File: day-04/main.rb
# Advent of Code 2015 - Day 4

INPUT_KEY = "ckczppom".freeze

require "digest"

def md5_with_prefix(input, prefix, number = 0)
  md5 = Digest::MD5.new

  loop do
    return number if md5.hexdigest("#{input}#{number}").start_with?(prefix)
    number += 1
  end
end

def main
  magic_one = md5_with_prefix(INPUT_KEY, "0" * 5)
  magic_two = md5_with_prefix(INPUT_KEY, "0" * 6, magic_one)

  puts "Part One: #{magic_one}" # magic number for MD5 with 5 leading zeros
  puts "Part Two: #{magic_two}" # ... and with 6 leading zeros
end

main if to_s == "main"

File: day-04/main.py
"""Advent of Code 2015 - Day 4."""

import hashlib
import itertools


def md5_with_prefix(input, prefix, start_with=0):
    """Determine first number that generates a hash with the given prefix."""
    md5_input = hashlib.md5(input)

    for number in itertools.count(start_with):
        md5 = md5_input.copy()
        md5.update(str(number).encode('ascii'))
        if md5.hexdigest().startswith(prefix):
            return number


def main():
    """Main entry point of puzzle solution."""
    INPUT_KEY = b'ckczppom'

    magic_one = md5_with_prefix(INPUT_KEY, '0' * 5)
    magic_two = md5_with_prefix(INPUT_KEY, '0' * 6, start_with=magic_one)

    print('Part One: {}'.format(magic_one))
    print('Part Two: {}'.format(magic_two))


if __name__ == '__main__':
    main()
